<a href="https://colab.research.google.com/github/nsriniva/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [131]:
### YOUR WORK HERE

# Using Jimmy Spagle's idea to compute column names
from re import compile as rcompile
names_text = !curl https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names

rex = rcompile('[\d]+\. ([a-z-]+):')
# Create a generator
matches  = (rex.search(line) for line in names_text)
col_names = [ match.group(1) for match in matches if match ]

print('\n', len(col_names), col_names)


 26 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


In [132]:
import pandas as pd
import numpy as np

autos_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
df = pd.read_csv(autos_url,header=None, names=col_names)

In [133]:
display(df)

df.shape

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3                 ?  alfa-romero  ...       21          27  13495
1            3                 ?  alfa-romero  ...       21          27  16500
2            1                 ?  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

(205, 26)

In [134]:
# The names file(https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names)
# states that missing attribure values are denoted by '?':
#8. Missing Attribute Values: (denoted by "?")
#   Attribute #:   Number of instances missing a value:
#   2.             41
#   6.             2
#   19.            4
#   20.            4
#   22.            2
#   23.            2
#   26.            4

# Replace '?' with np.NaN
df = df.replace('?', np.NaN)

# Verify that the null value counts match the missing info counts from the names file
df.isnull().sum()


symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [135]:
### YOUR WORK HERE
dfs = df.sample(10, random_state=30)
dfs

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [136]:
display(df['highway-mpg'].describe())
df['highway-mpg'].mean()

count    205.000000
mean      30.751220
std        6.886443
min       16.000000
25%       25.000000
50%       30.000000
75%       34.000000
max       54.000000
Name: highway-mpg, dtype: float64

30.75121951219512

In [137]:
dfs['highway-mpg'].describe()

count    10.000000
mean     32.800000
std       7.269418
min      18.000000
25%      30.500000
50%      33.000000
75%      38.000000
max      43.000000
Name: highway-mpg, dtype: float64

In [138]:
from scipy import stats

stats.ttest_1samp(dfs['highway-mpg'], popmean=30.75122)

Ttest_1sampResult(statistic=0.8912420269614587, pvalue=0.3960005156031261)

p-value is: The probability of the null being true, given the sample that we have collected

Confidence level: How likely does my alternative hypothesis have to be in order for me to reject the null = 95% = 0.95

P-value threshold: .05 (1-confidence level)

If P-value is <.05, I reject the null hypothesis

If P-Value >=.05 I fail to reject the null hypothesis

Conclusion: Based on a t-statistic of 0.8912 and a p-value of .3960, I fail to reject the null hypothesis that my sample's highway-mpg is reflective of the population.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [139]:
### YOUR WORK HERE
stats.ttest_1samp(dfs['highway-mpg'], popmean=35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Confidence level : 95% = 0.95
p-value threshold : 1 - 0.95 = 0.05

Conclusion: Based on a t-statistic of -0.9570 and a p-value of 0.3636, I fail to reject the null hypothesis that my sample's highway-mpg is representative of
the salesman's claim of 35 highway-mpg for the car lot

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [140]:
### YOUR WORK HERE
dfs2 = df.sample(100, random_state=30)
display(dfs2)
dfs2.shape

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
..         ...               ...            ...  ...      ...         ...    ...
189          3               NaN     volkswagen  ...       24          29  11595
192          0               NaN     volkswagen  ...       33          38  13845
171          2               134         toyota  ...       24          30  11549
203         -1                95          volvo  ...       26          27  22470
47           0               145         jaguar  ...       15          19  32250

[100 rows x 26 columns]

(100, 26)

In [141]:
res = stats.ttest_1samp(dfs2['highway-mpg'], popmean=35)

print(res.statistic, res.pvalue)

-6.5185242467032305 3.003137417468754e-09


Based on a t-statistic of -6.5185 and a p-value of 0.0000, I reject the null hypothesis that my sample's highway-mpg is representative of the salesman's claim of 35 highway-mpg for the car lot

With a sample size of 10 i could not reject the null hypothesis but with a sample size of 100, i can.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The different conclusions are due to the the size of the samples used for the t-tests - the larger sample size for the second sample is more representative of the dataset. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [142]:
### YOUR WORK HERE
names_text = !curl https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.names

rex = rcompile('[\d]+\. ([a-zA-Z -]+): 2')
# Create a generator
matches  = (rex.search(line) for line in names_text)
col_names = [ match.group(1) for match in matches if match ]

print('\n', len(col_names), col_names)


 17 ['Class Name', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']


In [143]:
house_votes_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'
hv_df = pd.read_csv(house_votes_url,header=None, names=col_names)

display(hv_df)

hv_df.shape

Class Name  ... export-administration-act-south-africa
0    republican  ...                                      y
1    republican  ...                                      ?
2      democrat  ...                                      n
3      democrat  ...                                      y
4      democrat  ...                                      y
..          ...  ...                                    ...
430  republican  ...                                      y
431    democrat  ...                                      y
432  republican  ...                                      y
433  republican  ...                                      y
434  republican  ...                                      n

[435 rows x 17 columns]

(435, 17)

In [144]:
hv_df = hv_df.replace('?',np.NaN)

hv_df.isnull().sum()

Class Name                                  0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue                     25
crime                                      17
duty-free-exports                          28
export-administration-act-south-africa    104
dtype: int64

In [145]:
hv_df = hv_df.replace('y',int(1)).replace('n', int(0))

display(hv_df)

display(hv_df.shape)

hv_df.isnull().sum()

Class Name  ...  export-administration-act-south-africa
0    republican  ...                                     1.0
1    republican  ...                                     NaN
2      democrat  ...                                     0.0
3      democrat  ...                                     1.0
4      democrat  ...                                     1.0
..          ...  ...                                     ...
430  republican  ...                                     1.0
431    democrat  ...                                     1.0
432  republican  ...                                     1.0
433  republican  ...                                     1.0
434  republican  ...                                     0.0

[435 rows x 17 columns]

(435, 17)

Class Name                                  0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue                     25
crime                                      17
duty-free-exports                          28
export-administration-act-south-africa    104
dtype: int64

In [146]:
grouped = hv_df.groupby(hv_df['Class Name'])

In [147]:
dem_df = grouped.get_group('democrat').drop(columns=['Class Name'])

display(dem_df)

display(dem_df.shape)

dem_df.isnull().sum()

handicapped-infants  ...  export-administration-act-south-africa
2                    NaN  ...                                     0.0
3                    0.0  ...                                     1.0
4                    1.0  ...                                     1.0
5                    0.0  ...                                     1.0
6                    0.0  ...                                     1.0
..                   ...  ...                                     ...
425                  0.0  ...                                     NaN
426                  1.0  ...                                     1.0
428                  NaN  ...                                     1.0
429                  1.0  ...                                     1.0
431                  0.0  ...                                     1.0

[267 rows x 16 columns]

(267, 16)

handicapped-infants                        9
water-project-cost-sharing                28
adoption-of-the-budget-resolution          7
physician-fee-freeze                       8
el-salvador-aid                           12
religious-groups-in-schools                9
anti-satellite-test-ban                    8
aid-to-nicaraguan-contras                  4
mx-missile                                19
immigration                                4
synfuels-corporation-cutback              12
education-spending                        18
superfund-right-to-sue                    15
crime                                     10
duty-free-exports                         16
export-administration-act-south-africa    82
dtype: int64

In [148]:
rep_df = grouped.get_group('republican').drop(columns=['Class Name'])

display(rep_df)

display(rep_df.shape)

rep_df.isnull().sum()

handicapped-infants  ...  export-administration-act-south-africa
0                    0.0  ...                                     1.0
1                    0.0  ...                                     NaN
7                    0.0  ...                                     1.0
8                    0.0  ...                                     1.0
10                   0.0  ...                                     0.0
..                   ...  ...                                     ...
427                  0.0  ...                                     1.0
430                  0.0  ...                                     1.0
432                  0.0  ...                                     1.0
433                  0.0  ...                                     1.0
434                  0.0  ...                                     0.0

[168 rows x 16 columns]

(168, 16)

handicapped-infants                        3
water-project-cost-sharing                20
adoption-of-the-budget-resolution          4
physician-fee-freeze                       3
el-salvador-aid                            3
religious-groups-in-schools                2
anti-satellite-test-ban                    6
aid-to-nicaraguan-contras                 11
mx-missile                                 3
immigration                                3
synfuels-corporation-cutback               9
education-spending                        13
superfund-right-to-sue                    10
crime                                      7
duty-free-exports                         12
export-administration-act-south-africa    22
dtype: int64

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [149]:
### YOUR WORK HERE
display(dem_df.describe())
display(rep_df.describe())

handicapped-infants  ...  export-administration-act-south-africa
count           258.000000  ...                              185.000000
mean              0.604651  ...                                0.935135
std               0.489876  ...                                0.246956
min               0.000000  ...                                0.000000
25%               0.000000  ...                                1.000000
50%               1.000000  ...                                1.000000
75%               1.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

handicapped-infants  ...  export-administration-act-south-africa
count           165.000000  ...                              146.000000
mean              0.187879  ...                                0.657534
std               0.391804  ...                                0.476168
min               0.000000  ...                                0.000000
25%               0.000000  ...                                0.000000
50%               0.000000  ...                                1.000000
75%               0.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

In [157]:
# The Null hypothesis is that Republicans and Democrats vote similarly on an 
# issue.
# The Alternate hypothesis is that Repulicans and Democrats vote differently on 
# an issue.
# We use a confidence level of 99% = 0.99
# => pvalue = 1 - 0.99 = 0.01
# If the ttest for an issue returns a pvalue < 0.01, then we can reject the 
# Null hypothesis i.e. We can say, with a 99% confidence level, that Democrats 
# and Repulicans vote differently on that issue. 


issues_diff = []
issues_sim = []
issues_tossup = []

for issue in dem_df.columns:
  res = stats.ttest_ind(dem_df[issue].notna(), rep_df[issue].notna())
  if res.pvalue < 0.01:
    issues_diff.append(issue)
  elif res.pvalue > 0.1:
    issues_sim.append(issue)
  else:
    issues_tossup.append(issue)

for issue in issues_diff:
  if dem_df[issue].notna().mean() > rep_df[issue].notna().mean():
    print(f'Democrats support {issue} more than Republicans\n')
  else:
    print(f'Republicans support {issue} more than Democrats\n')

if len(issues_sim) > 0:
  # These issues have a pvalue > 0.1 i.e. the Null hypothesis cannot be 
  # rejected and the difference is not significant at the 90% level.
  print('Republicans and Democrats vote similarly on:')
  display(issues_sim)



Democrats support aid-to-nicaraguan-contras more than Republicans

Republicans support export-administration-act-south-africa more than Democrats

Republicans and Democrats vote similarly on:


['handicapped-infants',
 'water-project-cost-sharing',
 'adoption-of-the-budget-resolution',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups-in-schools',
 'anti-satellite-test-ban',
 'immigration',
 'synfuels-corporation-cutback',
 'education-spending',
 'superfund-right-to-sue',
 'crime',
 'duty-free-exports']

In [158]:
issues_diff = []
issues_sim = []
issues_tossup = []

for issue in dem_df.columns:
  res = stats.ttest_ind(dem_df[issue], rep_df[issue], nan_policy='omit')
  if res.pvalue < 0.01:
    issues_diff.append(issue)
  elif res.pvalue > 0.1:
    issues_sim.append(issue)
  else:
    issues_tossup.append(issue)

for issue in issues_diff:
  if dem_df[issue].mean() > rep_df[issue].mean():
    print(f'Democrats support {issue} more than Republicans\n')
  else:
    print(f'Republicans support {issue} more than Democrats\n')

if len(issues_sim) > 0:
  # These issues have a pvalue > 0.1 i.e. the Null hypothesis cannot be 
  # rejected and the difference is not significant at the 90% level.
  print('Republicans and Democrats vote similarly on:')
  display(issues_sim)


Democrats support handicapped-infants more than Republicans

Democrats support adoption-of-the-budget-resolution more than Republicans

Republicans support physician-fee-freeze more than Democrats

Republicans support el-salvador-aid more than Democrats

Republicans support religious-groups-in-schools more than Democrats

Democrats support anti-satellite-test-ban more than Republicans

Democrats support aid-to-nicaraguan-contras more than Republicans

Democrats support mx-missile more than Republicans

Democrats support synfuels-corporation-cutback more than Republicans

Republicans support education-spending more than Democrats

Republicans support superfund-right-to-sue more than Democrats

Republicans support crime more than Democrats

Democrats support duty-free-exports more than Republicans

Democrats support export-administration-act-south-africa more than Republicans

Republicans and Democrats vote similarly on:


['water-project-cost-sharing']

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [151]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [152]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [153]:
### YOUR WORK HERE